File: using_python/90_subset2.py
# 输入: [1,2,2]
# 输出:
# [
#   [2],
#   [1],
#   [1,2,2],
#   [2,2],
#   [1,2],
#   []
# ]

from typing import List
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res =[[]]
        path =[]
        n = len(nums)
        def dfs(start,end,nums,path,res):
            for i in range(start,end):
                if i>start and nums[i]==nums[i-1]:
                    continue
                path.append(nums[i])
                res.append(path[:])
                dfs(i+1,end,nums,path,res)
                path.pop()


        dfs(0,n,nums,path,res)
        return res 


so =Solution()
print(so.subsetsWithDup([1,2,2]))

File: using_Go/90_subset2.go
package partition

import (
    "fmt"
)


func subsetsWithDup(nums []int) [][]int {
    n :=len(nums)
    sort.Ints(nums)
    path = make([]int,n)
    res =make([][]int,0)

    dfs(0,n,nums,&path,&res)
    return res 
}

func dfs (start int, end int, nums []int, path *[]int, res *[][]int ) {        
        for i:=start;i<end;i+=1{
            if (i>start && nums[i]==nums[i-1]){
                continue
            }
            path := append(path, nums[i])
            res := append(*res,path)
            path = path[:len(path)-1]
            dfs(i+1,end,nums,&path,&res)
        }
}

